diffkemp/snapshot.py

```diff
@@ -0,0 +1,238 @@
+        For sysctl option, glob_var is the data variable that the option sets.
+        """
+
+        def __init__(self, mod, glob_var, tag):
+            self.mod = mod
+            self.glob_var = glob_var
+            self.tag = tag
+
+    class FunctionGroup:
+        """
+        Group of functions. Used to group functions belonging to a single
+        sysctl options.
+        """
+
+        def __init__(self):
+            self.functions = dict()
+
+    def __init__(self, kernel_source=None, snapshot_source=None,
+                 fun_kind=None):
+        self.kernel_source = kernel_source
+        self.snapshot_source = snapshot_source
+        self.fun_kind = fun_kind
+        self.fun_groups = dict()
+        self.created_time = None
+
+    @classmethod
+    def create_snapshot_from_source(cls, kernel_dir, output_dir, fun_kind=None,
+                                    setup_dir=True):
+        """
+        Create a snapshot from a kernel source directory and prepare it for
+        snapshot directory generation.
+        :param kernel_dir: Source kernel directory.
+        :param output_dir: Snapshot output directory.
+        :param fun_kind: Snapshot function kind.
+        :param setup_dir: Whether to recreate the output directory.
+        :return: Desired instance of Snapshot.
+        """
+        output_path = os.path.abspath(output_dir)
+
+        # Cleanup or create the output directory of the new snapshot.
+        if setup_dir:
+            if os.path.isdir(output_path):
+                shutil.rmtree(output_path)
+            os.mkdir(output_path)
+
+        # Prepare source representations for the new snapshot
+        kernel_source = KernelSource(kernel_dir, True)
+        snapshot_source = KernelSource(output_path)
+
+        kernel_snapshot = cls(kernel_source, snapshot_source, fun_kind)
+
+        # Add a new None group to the snapshot if the function kind is None.
+        if fun_kind is None:
+            kernel_snapshot._add_none_group()
+
+        return kernel_snapshot
+
+    @classmethod
+    def load_snapshot_from_dir(cls, snapshot_dir, config_file="snapshot.yaml"):
+        """
+        Loads a snapshot from its directory.
+        :param snapshot_dir: Target snapshot directory.
+        :param config_file: Name of the snapshot configuration file.
+        :return: Desired instance of Snapshot.
+        """
+        snapshot_source = KernelSource(snapshot_dir)
+        loaded_snapshot = cls(None, snapshot_source)
+
+        with open(os.path.join(snapshot_dir, config_file), "r") as \
+                snapshot_yaml:
+            loaded_snapshot._from_yaml(snapshot_yaml.read())
+
+        return loaded_snapshot
+
+    def finalize(self):
+        """Finalize both kernel sources."""
+        if self.kernel_source is not None:
+            self.kernel_source.finalize()
+        if self.snapshot_source is not None:
+            self.snapshot_source.finalize()
+
+    def create_from_kernel(self):
```

Comment:
Maybe a better name would be `generate_files` or `create_snapshot_dir` or `generate_snapshot_dir`. The current name `create_from_kernel` seems a bit weird because there is no kernel as an argument to the call as I would expect from the name.